File: week-8/rimrim990/경주로 건설.py
from math import inf
from heapq import heappop, heappush

dy = [0,0,-1,1]
dx = [1,-1,0,0]

def solution(board):
  n = len(board)
  hq = [(0, 0, 0, 0), (0, 0, 0, 1)]
  visit = [[[inf, inf] for _ in range(n)] for _ in range(n)]
  visit[0][0] = [0, 0]

  while hq:
    acc, cy, cx, cdir = heappop(hq)
    for i in range(4):
      ny, nx = cy + dy[i], cx + dx[i]
      ndir = abs(dy[i])

      if 0 <= ny < n and 0 <= nx < n:
        if board[ny][nx] == 1:
          continue

        cost = acc + 100
        if cdir != ndir:
          cost += 500

        if cost < visit[ny][nx][ndir]:
          visit[ny][nx][ndir] = cost
          heappush(hq, (cost, ny, nx, ndir))

  return min(visit[n-1][n-1])

File: week-08/protoseo/경주로 건설.java
import java.util.*;

class Solution {
    int[] dx = {0, 1, 0, -1};
    int[] dy = {1, 0, -1, 0};

    public int solution(int[][] board) {
        int n = board.length;
        int[][][] isVisited = new int[n][n][4];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                Arrays.fill(isVisited[i][j], Integer.MAX_VALUE);
            }
        }

        Deque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{0, 0, 0});
        q.add(new int[]{0, 0, 1});
        for (int i = 0; i < 4; i++) {
            isVisited[0][0][i] = 0;
        }

        while (!q.isEmpty()) {
            int[] p = q.poll();
            int x = p[0];
            int y = p[1];
            int d = p[2];

            if (x == n && y == n) {
                continue;
            }

            for (int i = 0; i < 4; i++) {
                if (Math.abs(d - i) == 2) {
                    continue;
                }
                int xx = x + dx[i];
                int yy = y + dy[i];
                int cost = (i == d) ? 100 : 600;

                if (xx < 0 || yy < 0 || xx >= n || yy >= n) {
                    continue;
                }
                if (board[yy][xx] == 1 || isVisited[yy][xx][i] < isVisited[y][x][d] + cost) {
                    continue;
                }
                q.add(new int[]{xx, yy, i});
                isVisited[yy][xx][i] = isVisited[y][x][d] + cost;
            }
        }
        int answer = Integer.MAX_VALUE;
        for (int i = 0; i < 4; i++) {
            answer = Math.min(answer, isVisited[n - 1][n - 1][i]);
        }
        return answer;
    }
}
